Adding handler to collective quantization API

marjanfariborz opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #64573

Making the collective quantization API more flexible. This diff adds a handler that maps to each collective.

Differential Revision: [D30744267](https://our.internmc.facebook.com/intern/diff/D30744267/)

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang @cbalioglu @gcramer23

Commit:
Adding handler to collective quantization API

Making the collective quantization API more flexible. This diff adds a handler that maps to each collective.

Differential Revision: [D30744267](https://our.internmc.facebook.com/intern/diff/D30744267/)

[ghstack-poisoned]

Commit:
Update on "Adding handler to collective quantization API"


Making the collective quantization API more flexible. This diff adds a handler that maps to each collective.

Differential Revision: [D30744267](https://our.internmc.facebook.com/intern/diff/D30744267/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/algorithms/quantization/quantization.py:
```diff
@@ -109,32 +153,14 @@ def wrapper(*args, **kwargs):
+        in_splits = kwargs.get('in_splits', None)
+
+
+        if ((func == dist.all_to_all_single) or (func == dist.all_to_all) or (func == dist.all_gather)):
```
so that we don't have to update this line everytime a new API is added, might be better written as:


```
try:
    dispatch[func](...)
except KeyError:
    # no handler registered
    raise RuntimeError(...)
```

rohan-varma commented on torch/distributed/algorithms/quantization/quantization.py:
```diff
@@ -85,6 +83,52 @@ def _dequantize_tensor_list(tensor_list, qtype, quant_loss=None):
+    for i, t in enumerate(_dequantize_tensor_list(out_tensors, qtype, quant_loss=quant_loss)):
+        tensors[i] = t
+
+def _quantized_all_gather_handler(*args, **kwargs):
```
I feel like these 3 private functions can be written as a single one:

```
def _quantize_handler(quant_fn, dequant_fn, collective_fn, *args, **kwargs)
```

and you can pass in the `quant_fn`, `dequant_fn`, and `collective_fn` which are basically the 3 things that I can see distinguishing these functions in the implementation.

Once you have that, you can define names using `functools.partial`, for example:
```
_quantized_all_gather_handler = partial(_quantize_handler, _quantize_tensor_list, _dequantize_tensor_list, dist.all_to_all)
```

and so on. Finally you can update your `dispatch` dict to use these above names. The only tricky thing I see would be how to handle the output_splits and input_splits arguments in the a2a single handler.

rohan-varma reviewed (commented):
Great improvement to use a dispatcher for these APIs! Just a few comments for your consideration.

Commit:
Update on "Adding handler to collective quantization API"


Making the collective quantization API more flexible. This diff adds a handler that maps to each collective.

Differential Revision: [D30744267](https://our.internmc.facebook.com/intern/diff/D30744267/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/algorithms/quantization/quantization.py:
```diff
@@ -85,6 +83,52 @@ def _dequantize_tensor_list(tensor_list, qtype, quant_loss=None):
+    quant_loss = kwargs.get('quant_loss', None)
+    input_tensors = _quantize_tensor(input_tensor, qtype)
+    out_tensors = _quantize_tensor(tensors, qtype)
+    dist.all_to_all_single(out_tensors, input_tensors, out_splits, in_splits, group=group)
```
Maybe we can directly pass in kwargs here. Since `kwargs` is a dict, we can `pop` any keyword arguments that are not relevant to the distributed collective and then directly pass in the kwargs dict.

rohan-varma commented on torch/distributed/algorithms/quantization/quantization.py:
```diff
@@ -85,6 +83,52 @@ def _dequantize_tensor_list(tensor_list, qtype, quant_loss=None):
 
+def _quantized_a2a_single_handler(*args, **kwargs):
+    group = kwargs.get('group', None)
+    async_op = kwargs.get('async_op', False)
```
unused
